### PR TITLE
IVR registration add message receiver roles

### PR DESCRIPTION
**Pull request by NicolaasDieBaas**
change IVR flow to incorporate new states and logic for message receiver roles

### CODE REVIEW

**Comment by NicolaasDieBaas** [comment]
@gsvr please review

File: src/utils.js [hunk 1]
```diff
@@ -48,6 +48,8 @@ go.utils = {
 
     get_speech_option_days_time: function(days, time) {
         var speech_option;
+        console.log("DAYS: "+days);
+        console.log("TIME: "+time);
```
Comment (gsvr) [inline src/utils.js]: remove console.logs

File: src/utils.js [hunk 2]
```diff
@@ -514,6 +517,8 @@ go.utils = {
 
     save_contact_info_and_subscribe: function(im) {
         var mama_id = im.user.answers.mama_id;
+        console.log("SAVE CONTACT INFO AND SUBSCRIBE");
+        console.log("mama id: "+mama_id);
```
Comment (gsvr) [inline src/utils.js]: remove console.logs

File: src/voice_registration.js [hunk 3]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
+        // FreeText st-03
+        self.add('state_receiver_msisdn', function(name) {
+            var speech_option = 1;
+            return new FreeText(name, {
+                question: $('Please enter number'),
+                helper_metadata: go.utils.make_voice_helper_data(
+                    self.im, name, lang, speech_option),
+                next: function(content) {
+                    if (go.utils.is_valid_msisdn(content) === false) {
+                        return 'state_retry_personnel_auth';  // error message, allow retry
```
Comment (gsvr) [inline src/voice_registration.js]: I think this should be `state_retry_receiver_msisdn`

File: src/voice_registration.js [hunk 4]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
+        self.add('state_father_msisdn', function(name) {
+            var speech_option = '1';
+            return new FreeText(name, {
+                question: $('Please enter number (Father)'),
                 helper_metadata: go.utils.make_voice_helper_data(
                     self.im, name, lang, speech_option),
                 next: function(content) {
                     if (go.utils.is_valid_msisdn(content) === false) {
-                        return 'state_r02_retry_number';
+                        return 'state_retry_personnel_auth';  // error message, allow retry
```
Comment (gsvr) [inline src/voice_registration.js]: `state_retry_father_msisdn`?

File: src/voice_registration.js [hunk 5]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
             var speech_option = '1';
             return new FreeText(name, {
-                question: $('Retry number'),
+                question: $('Please enter number (Mother)'),
                 helper_metadata: go.utils.make_voice_helper_data(
                     self.im, name, lang, speech_option),
                 next: function(content) {
                     if (go.utils.is_valid_msisdn(content) === false) {
-                        return 'state_r02_retry_number';
+                        return 'state_retry_personnel_auth'; // error message, allow retry
```
Comment (gsvr) [inline src/voice_registration.js]: `state_retry_mother_msisdn`?

File: src/voice_registration.js [hunk 6]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
-            ];
-            if (self.im.user.answers.state_r04_mom_state === 'pregnant') {
-                choices = year_choices.slice(1,3);
-                speech_option = '1';
-            } else {
-                choices = year_choices.slice(0,2);
-                speech_option = '2';
-            }
+        // ChoiceState st-5A
+        self.add('state_5A_period_month', function(name) {
```
Comment (gsvr) [inline src/voice_registration.js]: Please replace the 5A here with `this_year` so it's clearer which state we're in.  Apply similarly to other state names.

File: src/voice_registration.js [hunk 7]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
+            var speech_option = 1;
+
             return new ChoiceState(name, {
-                question: $('Birth year?'),
+                question: $('Period month?'),
```
Comment (gsvr) [inline src/voice_registration.js]: Change this to `Period month this year?`  Apply similarly to other questions.

File: src/voice_registration.js [hunk 8]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
+                    new Choice('oct', $('October')),
+                    new Choice('nov', $('November')),
+                    new Choice('dec', $('December'))
+                ],
                 next: function(choice) {
-                    return 'state_r06_birth_month';
+                    var today = go.utils.get_today(self.im.config);
+                    var currentMonth = parseInt(today.format("MM"));
+                    var validStartMonth = currentMonth <= 10 ? 0 : currentMonth-10;
+                    var choiceMonth = parseInt(today.month(choice.value).format("MM"));
```
Comment (gsvr) [inline src/voice_registration.js:215]: This needs to move to a function and needs comments, but leave it for now if it works - open another issue to improve it.

File: src/voice_registration.js [hunk 9]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
+                question: $('Last period day {{ month }} [{{ year}}]'
+            ).context({ month: month, year: year }),
                 helper_metadata: go.utils.make_voice_helper_data(
                     self.im, name, lang, speech_option),
                 next: function(content) {
-                    var birth_date = go.utils.get_baby_dob(self.im, content);
-                    if (birth_date === 'invalid date') {
-                        return 'state_r14_retry_birth_day';
+                    var period_date = content+"-"+monthNum+"-"+year;
+                    if (!(new moment(period_date, "DD-MM-YYYY").isValid())) {
```
Comment (gsvr) [inline src/voice_registration.js]: Use `is_valid_date` function

File: src/voice_registration.js [hunk 10]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
+                        return 'state_retry_last_period_day';
                     } else {
-                        self.im.user.set_answer('birth_date', birth_date);
-                        return 'state_r09_language';
+                        self.im.user.set_answer('last_period_date', content+'-04-'+year);  //temp hard-coded
```
Comment (gsvr) [inline src/voice_registration.js]: use monthNum?

File: src/voice_registration.js [hunk 11]
```diff
@@ -34,191 +35,615 @@ go.app = function() {
...
+                question: $('Retry period day'
+            ).context({ month: month, year: year }),
+                helper_metadata: go.utils.make_voice_helper_data(
+                    self.im, name, lang, speech_option),
+                next: function(content) {
+                    var period_date = content+"-"+monthNum+"-"+year;
+                    if (!(new moment(period_date, "DD-MM-YYYY").isValid())) {
+                        return 'state_retry_last_period_day';
+                    } else {
+                        self.im.user.set_answer('last_period_date', content+'-04-'+year);  //temp hard-coded
```
Comment (gsvr) [inline src/voice_registration.js]: use monthNum?

File: src/voice_registration.js [hunk 12]
```diff
@@ -229,31 +654,41 @@ go.app = function() {
...
                     self.im, name, lang, speech_option),
                 choices: [
-                    new Choice('sms', $('sms')),
-                    new Choice('voice', $('voice'))
+                    new Choice('state_voice_days', $('voice')),
+                    new Choice('state_end_sms', $('sms'))
                 ],
                 next: function(choice) {
-                    return routing[choice.value];
+                    return choice.value;
```
Comment (gsvr) [inline src/voice_registration.js]: Please put back the `routing`, there's a reason for it.

File: src/voice_registration.js (hunk 8 above)
Comment (NicolaasDieBaas) [inline src/voice_registration.js:215]: created an issue for this

**Comment by NicolaasDieBaas** [comment]
@gsvr 2nd pass plz

**Comment by gsvr** [comment]
:+1:

**Pull request merged**